### ISSUE TITLE
CompendiumNG-portable.bat file is broken for me, by default

### ISSUE THREAD

**Issue by ghost** [CompendiumNG/CompendiumNG#221]
in this file: https://github.com/CompendiumNG/CompendiumNG/blob/2.1.5/application/src/main/resources/CompendiumNG-portable.bat
in order for it to work for me, the original becomes this:

``` bat
start "" /b "c:\Program Files\Java\jdk1.7.0_55\bin\javaw.exe" -Dlogback.configurationFile="C:\Program Files\CompendiumNG\logback.xml" -Dcompendiumng.config.dir="C:\Program Files\CompendiumNG\config" -Xmx512m -jar CompendiumNG.jar  %1 %2 %3 %4 %5 %6 %7 %8 %9
```

what changed:
- a) the path in `-Dcompendiumng.config.dir` must be quoted, otherwise nothing will happen/run and no errors are seen
- b) the right `javaw.exe` must be run, in my case the default one(in PATH) was java 1.6; (this might be complicated to get its path? but it's the same java (1.7 this time) that I used to run the installer - perhaps the installer can get the path to the javaw.exe that is used to run the installer but not by getting the parameter 0 of command line(if that even works for java programs) but rather by examining some (unknown to me at this time)java system property(like `java.version` and `java.home` are))
- c) the `start /b` should become `start "" /b` (where "" is the window title) because otherwise it complains about not finding `-Dlogback.configurationFile="C:\Program Files\CompendiumNG\logback.xml"` as if it tries to run it as a program - this is due to the path-ed `javaw.exe` being in double quotes and being taken(by `start`) as the window title instead of the command/program to run.

**Comment by maiklos** [comment]
- a) quotations are added now.
- b) regarding javaw.exe. It is run intentially this way so the OS pick up the default one which must found via PATH environment variable. We cannot hardwire absolute path to java since "Program Files" is a localized (maybe "Programme Files" on german system etc.). If you have java(and javaw) in PATH then you shouldn't need to explicitly enter full path to javaw.exe. 
- c) IMO this is not an issue if you don't put full path to the javaw.exe

M

**Comment by maiklos** [comment]
ad b) this http://www.oracle.com/technetwork/java/javase/runtime-windows-137915.html explains why you don't need to have full path to javaw on win system. java(x).exe is copied to windows system dir:

...
In addition, if no more recent version is already installed on the machine, the installer puts copies of the java and javaw application-launcher executables in the Microsoft Windows system directory. (The system directory varies depending on the Microsoft Windows version, but is usually either 
C:\winnt\system32 or C:\windows\system.) ...

(quoted from previously referenced link>)

so, if you run javaw and it doesn't start your default java then you should first make sure that default java is installed correctly and that it replaced java.exe and javaw.exe in your system directory

**Comment by ghost** [comment]
_TL;DR:_ when using JRE installer that article is spot on. When using JDK installer, one can have the choice of not installing java in windows system folders by deselecting `Public JRE` from the installer.

![public jre](https://cloud.githubusercontent.com/assets/5661825/3140787/5814e236-e93d-11e3-91d2-b53f381997fe.jpg)

--- long read:
I don't necessarily disagree. However that article seems to be incomplete. From my experience if you deselect `Public JRE` when installing, it will not install anything in the Windows system folders. In my case, I opted to have Public JRE when I installed the 1.6 version and deselected that when I installed jdk1.7 therefore what I have now is:

``` bat
C:\Users\user>where java.exe
C:\Windows\System32\java.exe

C:\Users\user>java -version
java version "1.6.0_30"
Java(TM) SE Runtime Environment (build 1.6.0_30-b12)
Java HotSpot(TM) 64-Bit Server VM (build 20.5-b03, mixed mode)

C:\Users\user>"c:\Program Files\Java\jdk1.7.0_55\bin\java.exe" -version
java version "1.7.0_55"
Java(TM) SE Runtime Environment (build 1.7.0_55-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.55-b03, mixed mode)
```

I had to start the installer with `"c:\Program Files\Java\jdk1.7.0_55\bin\javaw.exe" -jar CompendiumNG-installer.jar`  because otherwise it would say I am using java 1.6 and need java1.7  So I figured that maybe this way it can know the path to the java that is being used and pass it somehow to the .bat file.  

But the thing is, in Windows system folders there can be only one Public JRE installed: the latest one, which was selected by default on any java install, but if I don't select it, the previous one is left intact, even if it's outdated (which seems to be the opposite of what the article suggests).

I've just downloaded and installed jdk6u45 (jdk-6u45-windows-x64.exe) without selecting Public JRE to confirm that the outdated windir java is not overwritten. However while I've confirmed that is the case, I've just realized that maybe if I were to install just the JRE and not the JDK, then it would be... perhaps that article is true then for JRE installer(I notice now that it is called "Deploying the JRE on Windows" maybe there's one "Deploying the JDK on Windows" but I can't find it), but not for the JDK installer with Public JRE deselected. I'll test this right now. Ok, installing just the JRE (from jre-6u45-windows-x64.exe ) does replace the windir java:

``` bat
C:\Users\user>where java.exe
C:\Windows\System32\java.exe

C:\Users\user>java -version
java version "1.6.0_45"
Java(TM) SE Runtime Environment (build 1.6.0_45-b06)
Java HotSpot(TM) 64-Bit Server VM (build 20.45-b01, mixed mode)
```

To be honest I've never used the JRE installers, only the JDK ones - this is why I remember they allow me the option to not "poison" the windir java by deselecting the Public JRE. 

In a way, it would be much more consistent if the compendiumng installer would use the same java(in the .bat files) that it is used to start IT(it=the CompendiumNG-installer.jar), my case would be an example of why but also just in case someone has multiple javas in program files and is choosing one to use to start the installer - but then again, one could argue that if someone does that, then he can also edit the compendiumng batch files themselves and ensure the desired java is in use :) - don't mind me, this is just something that I would insist on doing if this were my project, I don't want to assert this on others: you do whatever you want:)  [sorry for this pathetic attempt at reverse psychology or something :D ]

Thanks. Peace.

**Comment by maiklos** [comment]
thanks thorough analysis. Frankly I also used to install only JDK version. I wasn't aware of java installer make copy of java(w).exe into the system directory (win only) until yesterday :-) - (note I primarily run linux)

CNG starter scripts relies on java(w).exe to be installed in any directory and the path to it be in the %PATH% so operating system can find it when you type java(w).exe. (approach pretty much consisten across operating systems). For CompendiumNG-portable.bat it is more important than for CompendiumNG.bat. With portable version you should be able to install CNG on i.e. USB flash drive, then take the USB flash drive and stick it into another computer (with java.exe installed possibly in another directory) and you should be able to run it without fiddling with the .bat (frankly extensive testing wasn't done in this area - especially to install on flash i.e. in win and then stick it into Linux/mac and use it there). If we resort to replace javaw.exe in the .bat file with absolute path then we will face other kind of problems.

In such cases like your, I would suggest to either:
a) install first java6 a then java7 (in order) - both with public JRE enabled
b) uninstall java6 and install only java7 (double check that java.exe is in the PATH)
c) don't install public JRE and edit PATH so in searched in the right directory
d) edit the resulting *.bat as you want - just bear in mind that with new version of CNG when you install it it will probably overwrite your *.bat and you will have to edit it again.

Actually you've risen a good point. Another user complained about not possibility to run CNG on win8 (I did clean installation on win8 and it worked without a glitch). So maybe the users is facing same problem with public JRE. Win with history of installations/deinstallations of JDK/JRE could easily end up in state like this with java6 public JRE in system to be inconsistent with the latest java one has on its' system.

Thanks
Peace ;-)

**Comment by ghost** [comment]
Oh you're right, I "forgot" we were talking about the portable bat ... definitely no good to have hardwired the path to java*.exe in that one... 
I agree with you now, to keep javaw as it is(pathless) in portable .bat

I was already on my way to choosing the e) solution which is: install manjaro linux(until I get enough info on how to install arch linux instead) and basically drop all windows(es) and anything proprietary like that. I'm kinda happy that I won't be using windows again, although I did like that UI consistency and speed (which is the opposited of what i've experienced with linux mint olivia 15 MATE ...)

Really appreciating all your input and thanks for your time on this. I guess it's safe to close this now.

Cheers
